### ISSUE TITLE
pallet-staking: era duration is off by one block.

### ISSUE THREAD

**Issue by thiolliere** [paritytech/substrate#7739]
In pallet-staking `end_era` is called through `pallet-session::end_session` during `on_initialize`. But `end_era` compute the duration of the era using the timestamp of the block, which is the timestamp of previous block due to timestamp inherent not being set yet for the current block. This means that the duration of the era is off doesn't take into account the duration of the last block of the era.

This should be fixed by computing the payout of the era in `on_finalize` when timestamp is set.

**Comment by kianenigma** [comment]
a demonstrative example of this: https://github.com/paritytech/substrate/pull/7719/files#diff-d78248c400ef5c7903736ec7b4a3103eb0be80c7c0f7472a4f8d43d01a583b2bR693-R711

**Comment by kianenigma** [comment]
Also, would be very good to document this diagram with this PR as well https://github.com/paritytech/substrate/pull/7719#discussion_r544296318

**Comment by kianenigma** [comment]
@thiolliere is this still relevant?

**Comment by thiolliere** [comment]
yes it isn't fix AFAIK, end_era is still triggered in on_initialize so it takes the block time of previous block.

But I don't think this issue is important.
